File: 程序员面试宝典/3.5.栈排序.py
class SortedStack:
    def __init__(self):
        self.help = []
        self.stack = []

    def push(self, val: int) -> None:
        while self.stack and val > self.stack[-1]:
            self.help.append(self.stack.pop(-1))

        self.stack.append(val)

        while self.help:
            self.stack.append(self.help.pop(-1))

    def pop(self) -> None:
        if self.stack:
            self.stack.pop(-1)

    def peek(self) -> int:
        if self.stack:
            return self.stack[-1]
        else:
            return -1

    def isEmpty(self) -> bool:
        return not self.stack

File: 程序员面试金典/3.5.栈排序.cpp
#include <stack>
using namespace std;

class SortedStack {
public:
    stack<int> st;
    stack<int> st1;

    SortedStack()
    {
    }

    void push(int val)
    {
        while (!st.empty() && st.top() < val) {
            st1.push(st.top());
            st.pop();
        }
        st.push(val);

        while (!st1.empty()) {
            st.push(st1.top());
            st1.pop();
        }
    }

    void pop()
    {
        if (!st.empty()) {
            st.pop();
        }
    }

    int peek()
    {
        if (!st.empty()) {
            return st.top();
        }
        return -1;
    }

    bool isEmpty()
    {
        return st.empty();
    }
};
